### PR TITLE
Column names order in "pivot_wider" with "glue_names"

### LINKED ISSUE #579: Incorrect order of colnames in `pivot_wider.()` with `names_glue`
`pivot_wider.()` assigns names to variables in incorrect order, when:

1. `names_glue` parameter is used with `{.value}` placeholder
2. variable is `names_from` is unsorted

Error is reproducible in both current CRAN version and development version.
See minimal example below. With `names_glue` (df2) columns are ordered as they would be without this option (df1), but column names are ordered as values appear in the lettr column. This issue is solved, when letter variable is sorted beforehand (df3).

``` r
set.seed(1)
src <- tidytable::tidytable(
  lettr = sample(letters, size = 5),
  v1 = 1:5
)

# correct
(df1 <- tidytable::pivot_wider.(
    src,
    names_from = lettr,
    values_from = v1
  ))
#> # A tidytable: 1 × 5
#>       a     b     d     g     y
#>   <int> <int> <int> <int> <int>
#> 1     4     5     2     3     1

# incorrect
(df2 <- tidytable::pivot_wider.(
  src,
  names_from = lettr,
  values_from = v1,
  names_glue = "{.value}_{lettr}"
))
#> # A tidytable: 1 × 5
#>    v1_y  v1_d  v1_g  v1_a  v1_b
#>   <int> <int> <int> <int> <int>
#> 1     4     5     2     3     1

# correct
(df2_2 <- tidytable::pivot_wider.(
  tidytable::arrange.(src, lettr),
  names_from = lettr,
  values_from = v1,
  names_glue = "{.value}_{lettr}"
))
#> # A tidytable: 1 × 5
#>    v1_a  v1_b  v1_d  v1_g  v1_y
#>   <int> <int> <int> <int> <int>
#> 1     4     5     2     3     1
```

<sup>Created on 2022-09-05 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

<details>
<summary>Session info</summary>

``` r
> sessioninfo::session_info()
─ Session info ────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
 setting  value
 version  R version 4.2.1 (2022-06-23 ucrt)
 os       Windows 10 x64 (build 22000)
 system   x86_64, mingw32
 ui       RStudio
 language (EN)
 collate  Russian_Russia.utf8
 ctype    Russian_Russia.utf8
 tz       Europe/Moscow
 date     2022-09-05
 rstudio  2022.07.0+548 Spotted Wakerobin (desktop)
 pandoc   2.18 @ C:/Program Files/RStudio/bin/quarto/bin/tools/ (via rmarkdown)

─ Packages ────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
 package     * version date (UTC) lib source
 callr         3.7.2   2022-08-22 [1] CRAN (R 4.2.1)
 cli           3.3.0   2022-04-25 [1] CRAN (R 4.2.0)
 clipr         0.8.0   2022-02-22 [1] CRAN (R 4.2.0)
 crayon        1.5.1   2022-03-26 [1] CRAN (R 4.2.0)
 curl          4.3.2   2021-06-23 [1] CRAN (R 4.2.0)
 data.table    1.14.2  2021-09-27 [1] CRAN (R 4.2.0)
 digest        0.6.29  2021-12-01 [1] CRAN (R 4.2.0)
 ellipsis      0.3.2   2021-04-29 [1] CRAN (R 4.2.0)
 evaluate      0.16    2022-08-09 [1] CRAN (R 4.2.1)
 fansi         1.0.3   2022-03-24 [1] CRAN (R 4.2.0)
 fastmap       1.1.0   2021-01-25 [1] CRAN (R 4.2.0)
 fs            1.5.2   2021-12-08 [1] CRAN (R 4.2.0)
 glue          1.6.2   2022-02-24 [1] CRAN (R 4.2.0)
 highr         0.9     2021-04-16 [1] CRAN (R 4.2.0)
 htmltools     0.5.3   2022-07-18 [1] CRAN (R 4.2.1)
 knitr         1.40    2022-08-24 [1] CRAN (R 4.2.1)
 lifecycle     1.0.1   2021-09-24 [1] CRAN (R 4.2.0)
 magrittr      2.0.3   2022-03-30 [1] CRAN (R 4.2.0)
 pillar        1.8.1   2022-08-19 [1] CRAN (R 4.2.1)
 pkgbuild      1.3.1   2021-12-20 [1] CRAN (R 4.2.0)
 pkgconfig     2.0.3   2019-09-22 [1] CRAN (R 4.2.0)
 prettyunits   1.1.1   2020-01-24 [1] CRAN (R 4.2.0)
 processx      3.7.0   2022-07-07 [1] CRAN (R 4.2.1)
 ps            1.7.1   2022-06-18 [1] CRAN (R 4.2.0)
 purrr         0.3.4   2020-04-17 [1] CRAN (R 4.2.0)
 R6            2.5.1   2021-08-19 [1] CRAN (R 4.2.0)
 remotes       2.4.2   2021-11-30 [1] CRAN (R 4.2.0)
 reprex        2.0.2   2022-08-17 [1] CRAN (R 4.2.1)
 rlang         1.0.5   2022-08-31 [1] CRAN (R 4.2.1)
 rmarkdown     2.16    2022-08-24 [1] CRAN (R 4.2.1)
 rprojroot     2.0.3   2022-04-02 [1] CRAN (R 4.2.0)
 rstudioapi    0.14    2022-08-22 [1] CRAN (R 4.2.1)
 sessioninfo   1.2.2   2021-12-06 [1] CRAN (R 4.2.0)
 tibble        3.1.8   2022-07-22 [1] CRAN (R 4.2.1)
 tidyselect    1.1.2   2022-02-21 [1] CRAN (R 4.2.0)
 tidytable     0.8.1.9 2022-09-05 [1] Github (markfairbanks/tidytable@5267e8a)
 utf8          1.2.2   2021-07-24 [1] CRAN (R 4.2.0)
 vctrs         0.4.1   2022-04-13 [1] CRAN (R 4.2.0)
 withr         2.5.0   2022-03-03 [1] CRAN (R 4.2.0)
 xfun          0.32    2022-08-10 [1] CRAN (R 4.2.1)
 yaml          2.3.5   2022-02-21 [1] CRAN (R 4.2.0)

 [1] C:/Users/Darx/AppData/Local/R/win-library/4.2
 [2] C:/Program Files/R/R-4.2.1/library

───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
```

</details>

### PR DESCRIPTION
**Pull request by Darxor**
Fixes #579 

Variable `glue_vars` now has names, that mimic `data.table::dcast`'s output column names
These names are used to map values onto new column names after `dcast` call

Tests should also probably be added to cover this edge case.

### CODE REVIEW

File: NEWS.md [hunk 1]
```diff
@@ -8,6 +8,7 @@
 #### Bug fixes
 * `tidytable::'%in%'` dispatches to `base::'%in%'` when comparing with a list (#563)
 * `pivot_wider.()`: Works with column names with spaces (#569)
+* `pivot_wider.()`: `names_glue="{.value}_{somecolumn}"` assigns column names in correct order (#579)
```
Comment (markfairbanks) [inline NEWS.md:11]: Want to put `(@Darxor, #579)` at the end so it's marked that you fixed it?

File: R/pivot_wider.R [hunk 2]
```diff
@@ -70,12 +70,7 @@ pivot_wider..tidytable <- function(.df,
   names_from <- tidyselect_names(.df, {{ names_from }})
   values_from <- tidyselect_names(.df, {{ values_from }})
 
-  uses_dot_value <- FALSE
-  if (!is.null(names_glue)) {
-    if (str_detect.(names_glue, ".value")) {
-      uses_dot_value <- TRUE
-    }
-  }
+  uses_dot_value <- !is.null(names_glue) && str_detect.(names_glue, ".value")
```
Comment (markfairbanks) [inline R/pivot_wider.R:73]: Yeah this is much better 😄

File: R/pivot_wider.R [hunk 3]
```diff
@@ -96,9 +91,16 @@ pivot_wider..tidytable <- function(.df,
...
     glue_df <- vec_rep(glue_df, length(values_from))
-    glue_df$.value <- vec_rep_each(values_from, values_from_reps)
+    glue_df <- mutate.(glue_df,
+                       .value = vec_rep_each(values_from, values_from_reps),
+                       .before = 1)
 
     glue_vars <- as.character(glue_data(glue_df, names_glue))
+    # mimic column names assigned by data.table::dcast()
+    if (length(values_from) <= 1) {
+      glue_df[[".value"]] <- NULL
```
Comment (markfairbanks) [inline R/pivot_wider.R:101]: Can you use one of these options instead of `glue_df[[".value"]] <- NULL`?

They both basically do the same thing but use the `data.table` way to remove columns. (Not that it matters here but they both avoid modify-by-reference as well.)

Option 1: `glue_df <- dt_j(glue_df, .value := NULL)`

Option 2: `glue_df <- mutate.(glue_df, .value = NULL)`

File: R/pivot_wider.R [hunk 4]
```diff
@@ -96,9 +91,16 @@ pivot_wider..tidytable <- function(.df,
...
+    }
+    names(glue_vars) <- do.call(paste, c(glue_df, list(sep = names_sep)))
```
Comment (markfairbanks) [inline R/pivot_wider.R:103]: Can we use `exec()` here? `do.call()` with extra arguments passed in a `list()` is a bit hard to read IMO.

``` r
names(glue_vars) <- exec(paste, !!!glue_df, sep = names_sep)
```

**Review by markfairbanks — CHANGES_REQUESTED** [review]

**Comment by markfairbanks** [comment]
This looks great @Darxor - thanks for the help! I have a few small changes requested above.

Can you also do me a favor and add this test to the bottom of `test-pivot_wider.R`?

``` r
test_that("correctly labels columns when `names_glue` is used, #570", {
  # length(values_from) == 1
  df1 <- tidytable(
    lettr = c("b", "a", "c"),
    v1 = c("b", "a", "c")
  )

  result1 <- pivot_wider.(
    df1,
    names_from = lettr,
    values_from = v1,
    names_glue = "{.value}_{lettr}"
  )

  expect_named(result1, c("v1_a", "v1_b", "v1_c"))
  expect_equal(unname(unlist(result1)), c("a", "b", "c"))

  # length(values_from) > 1
  df2 <- tidytable(
    lettr = c("b", "a", "c"),
    v1 = c("b", "a", "c"),
    v2 = c("b", "a", "c")
  )

  result2 <- pivot_wider.(
    df2,
    names_from = lettr,
    values_from = c(v1, v2),
    names_glue = "{.value}_{lettr}"
  )

  expect_named(result2, c("v1_a", "v1_b", "v1_c", "v2_a", "v2_b", "v2_c"))
  expect_equal(unname(unlist(result2)), c("a", "b", "c", "a", "b", "c"))
})
```

File: R/pivot_wider.R (hunk 4 above)
Comment (Darxor) [reply R/pivot_wider.R:103]: Agreed! I've tested it now, exec and unquoting add only negligible constant overhead of about 10-15µs.

**Comment by Darxor** [comment]
Thanks for your feedback @markfairbanks! All done now, happy to help :)

**Review by markfairbanks — APPROVED** [review]

**Comment by markfairbanks** [comment]
Thanks again! Would you like to be added to the contributor list? Basically your name would show up [here](https://cran.r-project.org/web/packages/tidytable/index.html) under the authors as a contributor.

**Comment by Darxor** [comment]
Absolutely! I would be quite honored to see my name there! :)

**Comment by markfairbanks** [comment]
Is "Alexander Sevostianov" correct for first and last name?

**Comment by Darxor** [comment]
Yup, it is
